torch/csrc/jit/interpreter.cpp

```diff
@@ -1,42 +1,208 @@
+  bool has_handle = hasHandleOutput(op);
+  return [=](const refcounted_list & inputs, refcounted_list & outputs) {
+    AutoGIL gil;
+    py::tuple py_inputs(op->cconv.size());
+    size_t i = 0;
+    size_t next_scalar = 0;
+    size_t next_tensor = 0;
+    HandleBuilder builder(has_handle);
+    for(auto arg_type : op->cconv) {
+      if(arg_type == 's') {
+        py_inputs[i] = py::reinterpret_borrow<py::object>(op->scalar_args[next_scalar++].get());
+      } else if(arg_type == 't') {
+        py_inputs[i] = THPVariable_Wrap(
+          builder.addInput(inputs.at(next_tensor), op->var_flags.at(next_tensor)));
+        next_tensor++;
+      }
+      i++;
+    }
+    py::object py_outputs(func(*py_inputs));
+
+    auto addOutput = [&](py::handle entry) {
+      if(!THPVariable_Check(entry.ptr())) {
+        throw std::runtime_error("Function.apply returned a non-Variable output");
+      }
+      THPVariable *var = (THPVariable*) entry.ptr();
+      outputs.push_back(builder.addOutput(var->cdata));
+    };
+    if(!PyTuple_Check(py_outputs.ptr())) {
+      addOutput(py_outputs);
+    } else {
+      for(py::handle entry : py::tuple(py_outputs)) {
+        addOutput(entry);
+      }
+    }
+    builder.writeTo(outputs);
+  };
+}
+
+Operation createCppCallback(CppOp* op) {
+  std::shared_ptr<autograd::Function> func = op->fn;
+  bool has_handle = hasHandleOutput(op);
+  return [=](const refcounted_list & inputs, refcounted_list & outputs) {
+    HandleBuilder builder(has_handle);
+    autograd::variable_list v_inputs;
+    for(size_t i = 0; i < inputs.size(); i++) {
+      v_inputs.push_back(builder.addInput(inputs[i], op->var_flags[i]));
+    }
+    autograd::variable_list v_outputs = func->apply(v_inputs);
+    for(auto & output : v_outputs) {
+      outputs.push_back(builder.addOutput(output));
+    }
+    builder.writeTo(outputs);
+  };
+}
+
+Operation createEvalCallback(CppOp * op) {
+  bool has_handle_output = hasHandleOutput(op);
+  return [=](const refcounted_list & inputs,
+             refcounted_list & outputs) {
+    Handle * handle_in = dynamic_cast<Handle*>(inputs.back());
+    JIT_ASSERT(handle_in);
+    HandleBuilder builder(has_handle_output);
+    auto& engine = autograd::python::PythonEngine::getDefaultEngine();
+    autograd::variable_list v_inputs;
+    for(size_t i = 0; i < inputs.size() - 1; i++) {
+      v_inputs.push_back(builder.addInput(inputs[i], op->var_flags[i]));
+    }
+    autograd::Engine::pre_callback_map callbacks;
+    callbacks.emplace(handle_in->forward_inputs.get(), [&](autograd::Function * _unused, autograd::variable_list & values) -> bool {
+      for(auto & v : values) {
+        outputs.push_back(builder.addOutput(v));
+      }
+      return false; // stop output and do not run DummyFunction
+    });
+    // node handle_in->use_count() == 1 means that we are guarenteed that we have the only
+    // only copy of the backward pass, but it is not clear that it is safe even in that case
```

Comment:
"it is not clear that it is safe even in that case" - can you make this more specific? I also can't see how `use_count() == 1` implies that we have "the only copy of backward pass" (especially that I don't know what a copy of backward pass is).

Reply:
So you have something like:

stage 1: handle0 -> Function0 -> Function1 -> Function2 -> Grad Accumulator
stage 2: handle1 -> Function3 -> Function 4 -> Function0 -> Grad Accumulator

Function 0 exists in both graphs because copied_next_fns told handle1 to inherit it from handle0 to calculate a derivative path that goes through the earlier stage.

Even if handle1.use_count() == 1, you can't free its graph unless handle0->use_count() == 0.
I think the 'right' solution here is for 'keep_graph=False' to actually mean "release the variables incrementally as this is run if we are guaranteed that the Function will be freed after the autograd finishes". But that condition is hard to check so is probably not worth trying to do.